src/controllers/hid/hidiothread.cpp

```diff
@@ -0,0 +1,330 @@
+#include "controllers/hid/hidiothread.h"
+
+#include <hidapi.h>
+
+#include <QTimer>
+
+#include "controllers/defs_controllers.h"
+#include "controllers/hid/legacyhidcontrollermappingfilehandler.h"
+#include "moc_hidiothread.cpp"
+#include "util/string.h"
+#include "util/time.h"
+#include "util/trace.h"
+
+namespace {
+constexpr int kReportIdSize = 1;
+constexpr int kMaxHidErrorMessageSize = 512;
+
+// Sleep time of run loop, in idle case, when no time consuming operation was executed before
+// The lower the time the more even the CPU load is spread over time
+// High values block the IO of this device and reduce response time and also the bandwidth.
+// Note, that this value has no influence on the amount of data to process,
+// the rate of InputReports is defined by the HID device itself.
+// This time should be below the rate of the HID device, which is ~1kHz for typical DJ controllers
+// the fastest possible rate of HID devices with USB HighSpeed or USB SuperSpeed interface is 8kHz
+constexpr int kSleepTimeWhenIdleMicros = 250;
+
+QString loggingCategoryPrefix(const QString& deviceName) {
+    return QStringLiteral("controller.") +
+            RuntimeLoggingCategory::removeInvalidCharsFromCategory(deviceName.toLower());
+}
+} // namespace
+
+HidIoThread::HidIoThread(
+        hid_device* pHidDevice, const mixxx::hid::DeviceInfo& deviceInfo)
+        : QThread(),
+          m_deviceInfo(deviceInfo),
+          // Defining RuntimeLoggingCategories locally in this thread improves runtime performance significiantly
+          m_logBase(loggingCategoryPrefix(deviceInfo.formatName())),
+          m_logInput(loggingCategoryPrefix(deviceInfo.formatName()) +
+                  QStringLiteral(".input")),
+          m_logOutput(loggingCategoryPrefix(deviceInfo.formatName()) +
+                  QStringLiteral(".output")),
+          m_pHidDevice(pHidDevice),
+          m_lastPollSize(0),
+          m_pollingBufferIndex(0),
+          m_runLoopSemaphore(1) {
+    // Initializing isn't strictly necessary but is good practice.
+    for (int i = 0; i < kNumBuffers; i++) {
+        memset(m_pPollData[i], 0, kBufferSize);
+    }
+    m_outputReportIterator = m_outputReports.begin();
+    m_state.storeRelease(static_cast<int>(HidIoThreadState::Initialized));
+}
+
+HidIoThread::~HidIoThread() {
+    hid_close(m_pHidDevice);
+}
+
+void HidIoThread::run() {
+    m_runLoopSemaphore.acquire();
```

Comment:
Please use QSemaphoreReleaser for this purpose and remove the manual release call when execiting this function.